overlord/fdestate/extop.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package fdestate
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// ExternalOperationStatus captures the status of operations running externally.
+type ExternalOperationStatus int
+
+// Admitted status values for changes and tasks.
+const (
+	DefaultStatus ExternalOperationStatus = 0
+
+	// Preparing means that we are performing preparation steps, but the
+	// operation isn't yet running externally.
+	PreparingStatus ExternalOperationStatus = 1
+
+	// DoingStatus means the operation is running externally.
+	DoingStatus ExternalOperationStatus = 2
+
+	// DoneStatus means the operation has completed successfully.
+	DoneStatus ExternalOperationStatus = 3
+
+	// CompletingStatus means the operation is completing.
+	CompletingStatus ExternalOperationStatus = 4
```

Comment:
does Completing comes before Done,  maybe would be good in the main comment to list the expected sequence(s) of statues?

Reply:
I'll add a small sequence diagram.